// @copilotkit/react-ui/dist/index.mjs
"use client";
import "./chunk-EFZPSZWO.mjs";
import "./chunk-MMVDU6DF.mjs";
import "./chunk-SC6JRFAJ.mjs";
import {
  CopilotSidebar
} from "./chunk-7RDGZ5JL.mjs";
import "./chunk-WB3YULQ4.mjs";
import {
  CopilotPopup
} from "./chunk-BAD2NFZ7.mjs";
import "./chunk-H5Y4W7DM.mjs";
import "./chunk-C3GSYRC3.mjs";
import "./chunk-GDSZGYCE.mjs";
import "./chunk-V7W6IM2V.mjs";
import {
  CopilotDevConsole
} from "./chunk-FFJHOZX6.mjs";
import "./chunk-Q5V6S67N.mjs";
import {
  shouldShowDevConsole
} from "./chunk-JY2CSDKN.mjs";
import "./chunk-KXE2JCUH.mjs";
import "./chunk-NRA3CFEE.mjs";
import "./chunk-BH6PCAAL.mjs";
import "./chunk-UFN2VWSR.mjs";
import {
  CopilotChat
} from "./chunk-ME2WM7IP.mjs";
import {
  Suggestions
} from "./chunk-226ZMOE3.mjs";
import {
  Suggestion
} from "./chunk-W26XFBEG.mjs";
import "./chunk-PLHTVHUW.mjs";
import "./chunk-EZQA3UPQ.mjs";
import "./chunk-ELGRNEAO.mjs";
import "./chunk-QIOJXTIQ.mjs";
import "./chunk-W7ONZTSW.mjs";
import "./chunk-2LIO4Z3E.mjs";
import "./chunk-5XLGXUQI.mjs";
import {
  AssistantMessage
} from "./chunk-3QYTKBWC.mjs";
import {
  ImageRenderer
} from "./chunk-DBKRAOH7.mjs";
import {
  UserMessage
} from "./chunk-VVL6JFCJ.mjs";
import {
  Markdown
} from "./chunk-JZ3RFQQ6.mjs";
import "./chunk-O72ZB5V3.mjs";
import {
  useChatContext
} from "./chunk-IEMQ2SQW.mjs";
import "./chunk-IK2BPURM.mjs";
import "./chunk-XWG3L6QC.mjs";
import "./chunk-IU3WTXLQ.mjs";
import "./chunk-T26KLXLH.mjs";
import {
  useCopilotChatSuggestions
} from "./chunk-EYRKZDP5.mjs";
import "./chunk-54JAUBUJ.mjs";
import "./chunk-JGMFJZMG.mjs";
import "./chunk-MRXNTQOX.mjs";
export {
  AssistantMessage,
  CopilotChat,
  CopilotDevConsole,
  CopilotPopup,
  CopilotSidebar,
  ImageRenderer,
  Markdown,
  Suggestion as RenderSuggestion,
  Suggestions as RenderSuggestionsList,
  UserMessage,
  shouldShowDevConsole,
  useChatContext,
  useCopilotChatSuggestions
};
